File: entrega_5/Java/src/topology/TopologyStar.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package topology;

import main_src.Particle;
import java.util.ArrayList;

/**
 *
 * @author Geraldo e Iago
 */
public class TopologyStar extends AbstractTopology
{
    private double globalBestFit;
    
    public TopologyStar() { }    
    
    @Override
    public void init(int size)
    {
        super.init(size);
        this.globalBestFit = Double.MAX_VALUE;//começa com o pior caso possível
    }
    
    @Override
    public void setGBest(double val, ArrayList<Particle> vec)
    {
        this.globalBestFit = val;
        Particle best = null;
        double fit = Double.MAX_VALUE;
        for(int i=0, size=vec.size(); i<size; i++)
        {
            Particle p = vec.get(i);
            if(p.getFitness() <= fit)
            {
                best = p;
                fit = p.getFitness();
            }
        }
        for(int i=0, size=vec.size(); i<size; i++)
        {
            Particle p = vec.get(i);
            p.setGBest(best.getData());
        }
    }
    
    @Override
    public double getBest()
    {
        return this.globalBestFit;
    }
}

File: entrega_1/pso_simples/topology/topologystar.cpp
#include "topologystar.h"

TopologyStar::TopologyStar():AbstractTopology()
{
    this->gBest = nullptr;
}

TopologyStar::~TopologyStar()
{

}

void TopologyStar::init(int size)
{
    AbstractTopology::init(size);
    this->globalBest = INT32_MAX;//pior caso possível
}

void TopologyStar::setBest(int pos, std::vector<Particle *> *vec)
{
    Particle *current = vec->at(pos);
    current->setGBest(gBest->getData());
}

void TopologyStar::updateBest(std::vector<Particle *> *vec)
{
    for(int i=0, size=vec->size(); i<size; ++i)
    {
        double fit = this->allFits->at(i);
        if(fit <= globalBest)
        {
            Particle* current = vec->at(i);
            globalBest = fit;
            this->gBest = current;
        }
    }
}

double TopologyStar::getBest()
{
    return this->globalBest;
    //pbest
    /*double out = 0;
    int size = gBest->getPBest()->size();
    for(int i=0; i<size; ++i)
    {
        out += gBest->getPBest()->at(i);
    }
    out /= size;
    return out;*/
}
